---
title: "Class_exercise"
author: "Georgia Barone"
date: "3/16/2023"
output: github_document
---

# Loading in the libraries needed for analysis
```{r setup, include=FALSE}
# packages & functions needed
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(tidyverse)
library(IRanges)
source("/scratch/Shares/rinnclass/CLASS_2023/geba9152/CLASS_2023/util/bchm_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/geba9152/CLASS_2023/util/my_class_functions.R")

# file paths
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/chipseqprofs"
peak_path <- "results/bwa/mergedLibrary/macs/broadPeak_good_files/"
broadpeakfilepath <- file.path(basepath, peak_path)
```

# Here I am starting to analyze my data for my proteins of interest:
# BCRA1, H3K36me3, MAX, JUN, POLR2A, POLR2A phospho S2, POLR2A phospho S5, & TBP
# First I will read in each replicate file
```{r load in peak files}
# import peaks
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)

# printing out a table of the number of peaks in each file:
peak_num <- sapply(peak_list, length) %>% as.data.frame()
# label column
names(peak_num) <- c("num_peaks")

# make dbp name a col.
peak_num <- peak_num %>%
  rownames_to_column(var = "dbp") %>%
  separate(col = dbp,  into = c('dbp', 'replicate'), sep = "_")

# set working directory
setwd("/scratch/Shares/rinnclass/CLASS_2023/geba9152/CLASS_2023/CLASSES/05_R_analyses/class_activity_03202023")

# saving to results folder
write_csv(peak_num, "results/num_peaks_df.csv")
```

# Now I am going to create consensus peaks for each protein
```{r consensus peaks}
# creating a list of unique dpbs 
dbps <- unique(sapply(names(peak_list), function(x) {
   unlist(strsplit(x, "_"))[1]
}))

# run function consensus_from_reduced
consensus_list <- lapply(dbps, consensus_from_reduced, peak_list)
names(consensus_list) <- dbps

# num_peaks for consensus 
num_consensus_peaks <- sapply(consensus_list, length) %>% 
  as.data.frame() %>%
  rownames_to_column( var = "dbp") %>%
  dplyr::rename(number_consensus_peaks = ".")

# merge into num_peaks
peak_num <- left_join(peak_num, num_consensus_peaks)

# export consensus peaks to results folder
write_csv(peak_num, "results/num_peaks_df.csv")

# setting file path to export
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/geba9152"
consensus_path <- "CLASS_2023/CLASSES/05_R_analyses/class_activity_03202023/consensus_peaks/"
exportpath <- file.path(basepath, consensus_path)

# export consensus_list as multiple .bed files
for(i in 1:length(consensus_list)) {
rtracklayer::export(consensus_list[[i]], paste0(exportpath, names(consensus_list)[i], "_consensus_peaks.bed") )}

# cleaning up .bed files

# making list of consensus .bed files
consensus_file_list <- list.files("/scratch/Shares/rinnclass/CLASS_2023/geba9152/CLASS_2023/CLASSES/05_R_analyses/class_activity_03202023/consensus_peaks", full.names = T, pattern = ".bed")

# lapply (for loop) across consensus file list to add colnames
# col names for .broadPeak are: chr, start, end, name, score, strand
peaks <- lapply(consensus_file_list, read.table, col.names = c("chr", "start", "end", "name", "score", "strand"))

# adding dpbs as names for the peaks file
names(peaks) <- dbps

# remove contigs from peak files
# make chromosomes of interest object
canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")

# use lapply with filter function to cannonical_chr
peaks <- lapply(peaks, function(x) x %>% filter(chr %in% canonical_chr))


# export new files without contigs
new_filenames <- paste0("consensus_peaks/", names(peaks), "_consensus.bed")

for(i in 1:length(peaks)) {
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}
```

# Now I am going to make my consensus peaks compatable with UCSC genome browser
```{r UCSC compatability}
# adding a header to make compatible with UCSC 
# using paste0 to print the header text & then adding the names as the value the header equals with 'names' function
headers <- paste0("track type=bed name=", names(peaks))

# create a path to export after we add header (using the loop below)
new_filenames <- paste0("ucsc_consensus_peaks/", names(peaks), ".bed")

# write new ucsc .bed files to ucsc file path using loop below
for(i in 1:length(peaks)) {
  # Write the header line
  writeLines(headers[[i]], new_filenames[[i]])
  # Append the broadPeak table data
  
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}
```

# Now I want to compare a protein with a previous analysis 
## Here my TBP consensus peak (from code above) overlayed with the Spring 2021 TBP chip data
## Shown in the JPG below are two consensus peaks called from my code, that align with the Spring 2021 data
```{r downloading in ucsc data}
# goto UCSC genome browser and load in a peak file for a given protein
# load in the data for the same protein from the previous analysis
# compare how your consensus peaks are similar or different to previous analyses

knitr::include_graphics("/scratch/Shares/rinnclass/CLASS_2023/geba9152/CLASS_2023/CLASSES/05_R_analyses/class_activity_03202023/TBP_genomebrowser.jpg") 

```

# Now I am going to determine how my peaks for each protein overlap annotations of the genome
# First I will find the overlaps between my consensus peaks with promoters of lncRNA and mRNA promoters
```{r overlap between consensus peaks with promoters of lncRNA and mRNA}
# find overlaps of promoters for each protein
# filepath to import peaks
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/geba9152"
peak_path <- "CLASS_2023/CLASSES/05_R_analyses/class_activity_03202023/consensus_peaks"
consensusPeakPath <- file.path(basepath, peak_path)

# making consensus peak file variable
consensus_peaks_files <- list.files(consensusPeakPath, 
                                             pattern = "*.bed",
                                             full.names = TRUE)

# lapply with import function to make a list of GRanges
consensus_peaks <- lapply(consensus_peaks_files, rtracklayer::import)

# adding DBP name to each list of GRanges from the file name
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2023/geba9152/CLASS_2023/CLASSES/05_R_analyses/class_activity_03202023/consensus_peaks/|_consensus.bed","", consensus_peaks_files)

# load in gencode lncrna & mrna annotation files 
load("/scratch/Shares/rinnclass/CLASS_2023/geba9152/CLASS_2023/CLASSES/05_R_analyses/01_peak_features/results/peak_features.RData")

# counting promoter overlaps
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, consensus_peaks, type = "counts")

# row sum for each DBP to get overlaps at each promoter
num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)

# filtering data frame down
num_peaks_df_for_analysis <- num_peaks_df %>% filter(!row_number() %in% c(2,3,5,6,8,9,11,12,14,15,17,18,20,21,23,24))

# adding % columns for all promoters
num_peaks_df_for_analysis$percent_peaks_overlap_promoters <- (num_peaks_df_for_analysis$peaks_overlapping_promoters/36814)*100
```
## results: 
#1) What can you determine from these overlaps?

### BRCA1 peak overlaps with promoters: 14420, 39.16% of all promoters
### H3K36me3 peak overlaps with promoters: 2179, 5.92% of all promoters
### JUN peak overlaps with promoters: 700, 1.90% of all promoters
### MAX peak overlaps with promoters: 18524, 50.32% of all promoters
### POL2RA peak overlaps with promoters: 5102, 13.86% of all promoters
### POL2RA phosphoS2 peak overlaps with promoters: 8411, 22.85% of all promoters
### POL2RA phosphoS5 peak overlaps with promoters: 17143, 46.57% of all promoters
### TBP peak overlaps with promoters: 8480, 23.03% of all promoters

### Note*** percentages obtained above are from dividing peaks at promoters by total peaks

# Now I want to compare the overlaps with lncRNA and mRNA promoters seperately 
```{r breaking up lncrna & mrna promoters}
# breaking into lncrna & mrna

# lncrna promoter overlaps
num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])
# mrna promoter overlaps
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])

# save to num_peaks_df
write_csv(num_peaks_df, "results/num_peaks_df.csv")

# adding % columns for mRNA promoters and lncRNA promoters
# mRNA promoters
num_peaks_df_for_analysis$percent_mRNApeaks_overlap_promoters <- (num_peaks_df_for_analysis$peaks_overlapping_mrna_promoters/num_peaks_df_for_analysis$num_peaks)*100

# lncRNA promoters
num_peaks_df_for_analysis$percent_lncRNApeaks_overlap_promoters <- (num_peaks_df_for_analysis$peaks_overlapping_lncrna_promoters/num_peaks_df_for_analysis$num_peaks)*100

# making df for mRNA promoters (just to visualize)
num_peaks_mRNA_lncRNA <- num_peaks_df_for_analysis[c(1,13,14,15)]
```
## results:
#1) What is the difference in overlaps between mRNA and lncRNA promoters

# mRNA promoters
### BRCA1 peak overlaps with mRNA promoters: 11032, 8.13% of all mRNA promoters
### H3K36me3 peak overlaps with mRNA promoters: 618, 0.18% of all mRNA promoters
### JUN peak overlaps with mRNA promoters: 469, 2.45% of all mRNA promoters
### MAX peak overlaps with mRNA promoters: 13353, 12.3% of all mRNA promoters
### POL2RA peak overlaps with mRNA promoters: 3854, 16.4% of all mRNA promoters
### POL2RA phosphoS2 peak overlaps with mRNA promoters: 57493, 3% of all mRNA promoters
### POL2RA phosphoS5 peak overlaps with mRNA promoters: 12730, 5.17% of all mRNA promoters
### TBP peak overlaps with mRNA promoters: 6724, 16.23% of all mRNA promoters

# lncRNA promoters
### BRCA1 peak overlaps with lncRNA promoters: 3388, 2.5% of all lncRNA promoters
### H3K36me3 peak overlaps with lncRNA promoters: 1561, 0.47% of all lncRNA promoters
### JUN peak overlaps with lncRNA promoters: 231, 1.2% of all lncRNA promoters
### MAX peak overlaps with lncRNA promoters: 5171, 4.76% of all lncRNA promoters
### POL2RA peak overlaps with lncRNA promoters: 1248, 5.31% of all lncRNA promoters
### POL2RA phosphoS2 peak overlaps with lncRNA promoters: 2662, 1.14% of all lncRNA promoters
### POL2RA phosphoS5 peak overlaps with lncRNA promoters: 4413, 1.8% of all lncRNA promoters
### TBP peak overlaps with lncrna promoters: 1756, 4.24% of all lncRNA promoters

### Note*** percentages obtained above are from dividing peaks at promoters by total peaks

# Proteins seem to have fewer peaks in lncRNA promoters compared to mRNA promoters
# This makes sense because mRNA transcripts produce proteins

# Now I am going to test if there is more binding over gene bodies than promoters
# I will seperate lncRNA and mRNA gene bodies to find the overlaps 

```{r extracting genebody overlaps counts data}

# finding overlaps with gene_bodies 
genebody_peak_counts <- count_peaks_per_feature(mrna_lncrna_genes, 
                                                consensus_peaks, 
                                                type = "counts")

# looking all gene bodies
num_peaks_df$peaks_overlapping_genebody <- 
  rowSums(genebody_peak_counts)

# lncRNA gene bodies 
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])

# mRNA gene bodies
num_peaks_df$peaks_overlapping_mrna_genebody <- 
  rowSums(genebody_peak_counts[,mrna_gene_ids])

# add to num_peaks_df for analysis
write_csv(num_peaks_df, "results/num_peaks_df.csv")

# adding % columns for total, mRNA, & lncRNA genebody overlaps

# genebody overlaps
num_peaks_df_for_analysis$percent_peaks_overlap_genebody <- (num_peaks_df_for_analysis$peaks_overlapping_genebody/num_peaks_df_for_analysis$num_peaks)*100

# mRNA genebody overlaps
num_peaks_df_for_analysis$percent_mRNApeaks_overlap_genebody <- (num_peaks_df_for_analysis$peaks_overlapping_mrna_genebody/num_peaks_df_for_analysis$num_peaks)*100

# lncRNA genebody overlaps
num_peaks_df_for_analysis$percent_lncRNApeaks_overlap_genebody <- (num_peaks_df_for_analysis$peaks_overlapping_lncrna_genebody/num_peaks_df_for_analysis$num_peaks)*100

# making df for mRNA promoters (just to visualize)
num_peaks_mRNA_lncRNA <- num_peaks_df_for_analysis[c(1,9,16,17,18)]
```
## results: 
# 1) Do my proteins have more overlaps with promoters or genebodies?

# all gene body overlaps
### BCRA1: 26993, 19.88% of all gene bodies
### H3K36me3: 76443, 22.81% of all gene bodies
### JUN: 3328, 17.36% of all gene bodies
### MAX: 60723, 55.95% of all gene bodies
### POL2RA: 7087, 30.16% of all gene bodies
### POL2RA phosphoS2: 42324, 22.15 of all gene bodies
### POL2RA phosphoS5: 53738, 21.87% of all gene bodies
### TBP: 11110, 26.81% of all gene bodies

# mRNA
### BCRA1 promoter overlaps (from above): 11032, 8.13% of all mRNA promoters
### BCRA1 gene body overlaps: 21406, 15.77% of mRNA gene bodies
## BCRA1 contains more gene body overlaps than promoter overlaps
### H3K36me3 promoter overlaps: 618, 0.18% of all mRNA promoters
### H3K36me3 gene body overlaps: 70974, 21.19% of mRNA gene bodies
## H3K36me3 contains more gene body overlaps than promoter overlaps
### JUN promoter overlaps: 2.45% of all mRNA promoters
### JUN gene body overlaps: 2486, 12.96% of mRNA gene bodies
## JUN contains more gene body overlaps than promoter overlaps
### MAX promoter overlaps: 12.3% of all mRNA promoters
### MAX gene body overlaps: 47064, 43.37% of mRNA gene bodies
## MAX contains more gene body overlaps than promoter overlaps
### POL2RA promoter overlaps: 16.4% of all mRNA promoters
### POL2RA gene body overlaps: 5656, 24.07% of mRNA gene bodies
## POL2RA contains more gene body overlaps than promoter overlaps
### POL2RA phosphoS2 promoter overlaps: 3% of all mRNA promoters
### POL2RA phosphoS2 gene body overlaps: 37283, 19.51% of mRNA gene bodies
## POL2RA phosphoS2 contains more gene body overlaps than promoter overlaps
### POL2RA phosphoS5 promoter overlaps: 5.17% of all mRNA promoters
### POL2RA phosphoS5 gene body overlaps: 45575, 18.54% of mRNA gene bodies
## POL2RA phosphoS5 contains more gene body overlaps than promoter overlaps
### TBP promoter overlaps: 16.23% of all mRNA promoters
### TBP gene body overlaps: 9040, 21.81% of mRNA gene bodies
## TBP contains more gene body overlaps than promoter overlaps

# lncRNA 
### BCRA1 promoter overlaps (from above): 2.5% of all lncRNA promoters
### BCRA1 gene body overlaps: 5587, 4.11% of lncRNA gene bodies
## BCRA1 contains more gene body overlaps than promoter overlaps
### H3K36me3 promoter overlaps: 0.47% of all lncRNA promoters
### H3K36me3 gene body overlaps: 5469, 1.63% of lncRNA gene bodies
## H3K36me3 contains more gene body overlaps than promoter overlaps
### JUN promoter overlaps: 1.2% of all lncRNA promoters
### JUN gene body overlaps: 842, 4.39% of lncRNA gene bodies
## JUN contains more gene body overlaps than promoter overlaps
### MAX promoter overlaps: 4.76% of all lncRNA promoters
### MAX gene body overlaps: 13659, 12.59% of lncRNA gene bodies
## MAX contains more gene body overlaps than promoter overlaps
### POL2RA promoter overlaps: 5.31% of all lncRNA promoters
### POL2RA gene body overlaps: 1431, 6.09% of lncRNA gene bodies
## POL2RA contains more gene body overlaps than promoter overlaps
### POL2RA phosphoS2 promoter overlaps: 1.14% of all lncRNA promoters
### POL2RA phosphoS2 gene body overlaps: 5041, 2.64% of lncRNA gene bodies
## POL2RA phosphoS2 contains more gene body overlaps than promoter overlaps
### POL2RA phosphoS5 promoter overlaps: 1.8% of all lncRNA promoters
### POL2RA phosphoS5 gene body overlaps: 8163, 3.32% of lncRNA gene bodies
## POL2RA phosphoS5 contains more gene body overlaps than promoter overlaps
### TBP promoter overlaps: 4.24% of all lncRNA promoters
### TBP gene body overlaps: 2070, 5% of lncRNA gene bodies
## TBP contains more gene body overlaps than promoter overlaps

### Note*** percentages obtained above are from dividing peaks at gene bodies by total peaks

# I am curious if my proteins are transcription factors so I will use the annotations in a cell paper I found and see
## Using this code, num_peaks will now contain a column that says wether my protein of interest is a TF 
```{r TF annotations}

# downloading TF annotations 
url <- "https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx"

destination_for_url <- "results/TF_annotations.xlsx"

# to download we can use download.file
download.file(url, destination_for_url)

# creating human TFs variable
human_tfs <- readxl::read_excel("results/TF_annotations.xlsx",
                                sheet = 2, skip = 1)

# rename the 4th column to indicate if it is a TF
names(human_tfs)[4] <- "is_tf"

# intersection of gene names that are in our ChIP data and has a TF identity
length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs$Name)))

# grabbing first 4 columns 
human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]

# adding new column names
names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

# merge TF annotation to num_peaks_df
num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

# Let's check how many NAs -- we should have some missing values
dim(num_peaks_df[is.na(num_peaks_df$tf),])

# cleaning up df 
num_peaks_df <- num_peaks_df[,1:12]
write_csv(num_peaks_df, "results/num_peaks_df.csv")
```

# It is nice and all to find overlaps, but I am interested in how many proteins bind a specific promoter.
# I will use my handy "occurence" parameter in " count peaks per feature" 
```{r peak occurence}

# using using count_peaks_per_feature to get occurence parameter
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, consensus_peaks, 
                                               type = "occurrence")

# double check that all lncrna & mrna genes are accounted for:
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# saving information to table
write.table(promoter_peak_occurence, "results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

# making sure promoter_peak_occurrence and lncrna_mrna_promoters are in the same order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# creating peak_occurence df
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "1kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))
```
## results: values found in the peak_occurence_df

### BCRA1: N/A - not in annotation file
### H3K36me3: N/A - not in annotation file
### JUN: max # of proteins on a promoter is 18
### MAX: max # of proteins on a promoter is 15
### POL2RA: N/A - not in annotation file
### POL2RA phosphoS2: N/A - not in annotation file
### POL2RA phosphoS5: N/A - not in annotation file
### TBP: max # of proteins on a promoter is 18

# Now I want to start plotting my results
# First I will see if there is a realtionship between peak number and total DNA covered
## Result:
### Yes there is a relationship between peak number and total DNA covered, the more number of peaks, the more DNA covered. 
### See below:
```{r}
ggplot(num_peaks_df, aes(x = num_peaks, 
                         y = total_peak_length)) +
  geom_point() 
```

# Now I want to color my plot by wether the protein is a TF or not.
### See below:
```{r}
ggplot(num_peaks_df, aes(x = num_peaks, 
                 y = total_peak_length,
                 color = tf)) +
  facet_wrap(tf ~ .) +
  geom_point() 
```

# I want to make a histogram of the number of peaks for each of my proteins
### See below:
```{r}
ggplot(num_peaks_df, aes(x = num_peaks)) +
  geom_histogram()
```

# Now I want to facet this by the type of DNA binding domain my protein has.
### See below:
```{r}
ggplot(num_peaks_df, aes(x = num_peaks, fill = tf)) +
   geom_histogram(bins = 30, position = "dodge")
```
